### PR TITLE
Shylu: allow fastilu to work with block CRS matrices

### PR DESCRIPTION
**Pull request by jgfouca**
@trilinos/shylu 
@trilinos/tpetra 

Also, add some BlockCRS helper functions to Tpetra

## Motivation

A customer wants to be able to use shylu::fastilu with blockCrs matrices

## Testing

Correctness:

1) Enable determinism and fine-granularity fills (FASTILU_ONE_TO_ONE_UNBLOCKED). Then run soak tests on both
   CPU and GPU machines with 2 fill levels. Specifics:
   * for N = 4, 6, 8, 9
     * for block_size of each int that evenly divides N
       * for nnz % 10, 20, 50, 100
         * Generate 10 random matrices with these properties
           * For each random matrix:
             * Run Ifpack_tif_belos with these inputs
               * Run blockcrs, crs, and original (a copy of original fastilu source) FastILU impls
                 * All class members (A, L, U) should uncompress to equivalent matrices
                 * Didn't worry about convergence details

2) Run valgrind with some of these

3) Repeat the basic 6w test a bunch of times, both CPU and GPU

Performance:

For 10000x10000, 100 nnz per row, 4x4 blocks, block dense

```
CPU:
JGF initialization of blocked took: 0.0351184
JGF initialization of unblocked took: 0.805753
JGF initialization of orig took: 0.820278
 Time Init: 4.16738
JGF compute of blocked took: 0.0162119
JGF compute of unblocked took: 0.0473241
JGF compute of orig took: 0.0492717
 Finished computing Ifpack2 preconditioner
  Time (s): 0.11435
JGF apply of blocked took: 0.00204011
JGF apply of unblocked took: 0.00214537
JGF apply of orig took: 0.00173839

GPU:
JGF initialization of blocked took: 0.0791826
JGF initialization of unblocked took: 1.95681
JGF initialization of orig took: 1.93094
 Time Init: 8.77923
JGF compute of blocked took: 0.003213
JGF compute of unblocked took: 0.00254731
JGF compute of orig took: 0.00244249
 Finished computing Ifpack2 preconditioner
  Time (s): 0.0123409
JGF apply of blocked took: 0.00140019
JGF apply of unblocked took: 0.000446767
JGF apply of orig took: 0.000458988
```

### CODE REVIEW

File: packages/shylu/shylu_dd/frosch/src/Tools/FROSch_Output.h [hunk 1]
```diff
@@ -55,11 +55,11 @@
         const bool throw_exception = !(COND); \
         if(throw_exception) { \
             Teuchos::TestForException_incrThrowNumber(); \
-            const int throwNumber = throwNumber; \
+            const int throwNumber = Teuchos::TestForException_getThrowNumber(); \
```
Comment (jgfouca) [inline packages/shylu/shylu_dd/frosch/src/Tools/FROSch_Output.h:58]: This was unrelated to fastilu, but was generating tons of warnings, so I went ahead and fixed it (I think this is the right fix).

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 2]
```diff
@@ -64,13 +64,172 @@
...
+// comment out next line to disable
 //#define FASTILU_DEBUG_OUTPUT
 
-//forward declaration
+// whether to print timings
+//#define FASTILU_TIMER
+
+// Whether to try to maintain exact behavior with unblocked impl. Comes at
+// a steep performance cost.
+//#define FASTILU_ONE_TO_ONE_UNBLOCKED
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:83]: This was the preprocessor setting I used for debugging. It forces fastilu to produce deterministic results such that block and unblocked impls produce the same result.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 3]
```diff
@@ -64,13 +64,172 @@
...
+    for (LO row_nnz = row_map(row); row_nnz < row_map(row+1); ++row_nnz) {
+      row_idx(nnz++) = row;
+    }
+  }
+  assert(nnz == row_map(nrows_blocked));
+}
+#endif
+
+// some useful preprocessor functions
+#ifdef FASTILU_TIMER
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:209]: Tried to encapsulate some things in macros to make the code a bit cleaner. This is unrelated to blockcrs.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 4]
```diff
@@ -115,19 +271,20 @@ class MemoryPrimeFunctorNnzCoo;
 template<class Ordinal, class Scalar, class ExecSpace>
 class MemoryPrimeFunctorNnzCsr;
 
-template<class Ordinal, class Scalar, class ExecSpace>
+template<class Ordinal, class Scalar, class ExecSpace, bool BlockCrsEnabled=false>
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:274]: Decided to use a template to enable the blockcrs capability. I didn't want to deal with performance issues related to runtime checks.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 5]
```diff
@@ -140,7 +297,337 @@ class FastILUPrec
         using STS = Kokkos::ArithTraits<Scalar>;
         using RTS = Kokkos::ArithTraits<Real>;
 
-    private:
+  struct IdentityFunctor
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:300]: A bunch of functors for when blockcrs has to do different thing than regular crs.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 6]
```diff
@@ -249,68 +737,13 @@ class FastILUPrec
         KernelHandle khL;
         KernelHandle khU;
 
-        //Internal functions
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:252]: This was dead code (not used anywhere).

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 7]
```diff
@@ -528,145 +935,103 @@ class FastILUPrec
...
+        {
+            using WithoutInit = Kokkos::ViewAllocateWithoutInitializing;
+
+            // Ensure all filled entries have the sentinel value
+#ifdef FASTILU_ONE_TO_ONE_UNBLOCKED
+            aVal_ = ScalarArrayMirror("aVal", aColIdx_.extent(0));
+#else
+            aVal_ = ScalarArrayMirror("aVal", aColIdx_.extent(0) * blockCrsSize * blockCrsSize);
+#endif
+            Kokkos::deep_copy(aVal_, std::numeric_limits<Scalar>::min());
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:989]: Min is used to denote values that FastILUFunctor should not skip (the fills). FastILUFunctor will skip true zeroes.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 8]
```diff
@@ -859,62 +1212,22 @@ class FastILUPrec
...
-            Kokkos::Timer Timer;
-            #endif
-#if 1
+            FASTILU_CREATE_TIMER(Timer);
             int nnzU = a2uMap.extent(0);
-            #if 1
-            ParPermCopyFunctor<Ordinal, Scalar, ExecSpace> permCopy(a2uMap, aVal, aRowIdx, uVal, uColIdx);
+            ParPermCopyFunctor<Ordinal, Scalar, ExecSpace, BlockCrsEnabled> permCopy(a2uMap, aVal, aRowIdx, aColIdx, uVal, uColIdx, blockCrsSize);
             Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpace>(0, nnzU), permCopy);
-            #else
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:883]: Dead code

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 9]
```diff
@@ -967,74 +1276,38 @@ class FastILUPrec
             diagElemsOut = diagElems;
         }
 
-        void applyDiagonalScaling()
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:970]: Dead code

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 10]
```diff
@@ -1184,137 +1462,147 @@ class FastILUPrec
...
-            
+
 
             // ------------------------------------------------
             // functor to load values
             // both matrices are sorted and, "a" (with fills) contains "aIn" (original)
             KOKKOS_INLINE_FUNCTION
             void operator()(const CopySortedValsTag &, const int i) const {
-                Ordinal aPtr = aRowMapIn[i];
+                Ordinal aPtr    = aRowMapIn[i];
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:1522]: There were bugs in the original impl where it was performing operations on the wrong row. The aPtrEnd check will prevent those and I added that check to several functors.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 11]
```diff
@@ -1587,59 +1802,64 @@ class FastILUPrec
...
             }
-            //Ordinal extent = aRowMap[nRows];
-            //ExecSpace().fence();
 
-            for (int i = 0; i < nFact; i++) 
+            for (int i = 0; i < nFact; i++)
             {
-                Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpace>(0, extent), iluFunctor);
+#ifdef FASTILU_ONE_TO_ONE_UNBLOCKED
+              // Force serialization to avoid races but still perform operation on device
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:1827]: I confirmed the race in the original impl. I think this is similar to par_ilut where the race is part of the algorithm.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 12]
```diff
@@ -1730,48 +1941,136 @@ class FastILUPrec
...
-            std::cout << "  >> compute done " << Timer2.seconds() << " <<" << std::endl << std::endl;
-            #endif
-            ExecSpace().fence();  //Fence so computeTime is accurate
+            ExecSpace().fence(); // Fence so computeTime is accurate
             computeTime = timer.seconds();
+            FASTILU_REPORT_TIMER(timer, "  >> compute done\n");
+        }
+
+        template <typename CRS>
+        void sptrsv_impl(ScalarArray &x, ScalarArray &y, CRS& crsmatL, CRS& crsmatU)
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:1955]: Factored this out to make it easier to work with both CRS and BlockCRS.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 13]
```diff
@@ -2327,12 +2567,96 @@ class FastILUFunctor
                         _Ux[uptr-1] = ((one - _omega) * _Ux[uptr - 1]) + (_omega * val);
                     }
                 }
+        }
+
+        KOKKOS_INLINE_FUNCTION
+        void functor_bcrs_impl(const Ordinal start, const Ordinal end) const
```
Comment (jgfouca) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:2573]: This functor was, by far, the biggest struggle. Getting something that matched the exact behavior of the unblocked impl was tricky. I tried to keep it as similar to the unblocked impl as possible, not sure if a different impl that took advantage of blocks more would be better.

**Review by jgfouca — COMMENTED** [review]
Annotations.

File: packages/tpetra/core/src/Tpetra_BlockCrsMatrix_Helpers_def.hpp
Comment (lucbv) [inline packages/tpetra/core/src/Tpetra_BlockCrsMatrix_Helpers_def.hpp:1]: A general description of both functions would be nice, basically just stating what the inputs are, what assumptions are made and what output can be expected

**Review by lucbv — CHANGES_REQUESTED** [review]

File: packages/tpetra/core/src/Tpetra_BlockCrsMatrix_Helpers_def.hpp
Comment (jgfouca) [reply packages/tpetra/core/src/Tpetra_BlockCrsMatrix_Helpers_def.hpp:1]: @lucbv , the descriptions are in the decl.

**Comment by jgfouca** [comment]
@lucbv , I think I addressed your modification requests.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp (hunk 13 above)
Comment (srajama1) [reply packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:2573]: I would have gone with the later, you went with the former, that is ok. At some point we might want to implement the later and compare.

**Review by srajama1 — DISMISSED** [review]
@jgfouca Some minor things. Otherwise this is fine.

**Review by lucbv — DISMISSED** [review]
Thanks for the small fixes

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 14]
```diff
@@ -528,145 +935,103 @@ class FastILUPrec
...
+            FASTILU_REPORT_TIMER(timer, " Copy time");
+
             // sort based on ColIdx, RowIdx stays the same (do we need this?)
             using host_space = typename HostSpace::execution_space;
             KokkosSparse::sort_crs_graph<host_space, OrdinalArrayMirror, OrdinalArrayMirror>
               (aRowMap_, aColIdx_);
-            #ifdef FASTILU_INIT_TIMER
-            host_space().fence();
-            std::cout << " Sort time : " << timer.seconds() << std::endl;
-            timer.reset();
```
[the remark this answers is not in the document]
Comment (jgfouca) [reply packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:542]: @srajama1 , I moved this code to `symbolicILU_common`. I noticed there was nearly identical code at the end of both versions of symbolicILU, so I encapsulated it in a new function.

**Review by srajama1 — APPROVED** [review]
